File: packager/packager.py
#!/usr/bin/env python

# this program takes in "unformatted" scroll data
# and turns it into a .volpkg
# a .volpkg has these properties
#  - 1 config file (config.json)
#    - number of slices
#    - dimensions
#    - 
#  - 1 folder containing scroll files in .tiff format
#    - the scroll naming convention will be the slices number
#  - 1 folder containing the point cloud file

import sys
import os
import subprocess
import shutil
import re
import json
import time

sys.stdout.write('vc_packager\n\n')

# check for help
if '-h' in sys.argv or '--help' in sys.argv:
    print \
"""[USAGE] vc_packager [-sp | --slices-path] <path to slices> [-of | --output-file] <name of output file>

[OPTIONS] 
[-sp | --slice-path] <path to slices>
    this option allows you to specify a custom path to slices
    the default is the current directory the program was run in

[-of | --output-file] <name of output file>
    this option allows you to specify the location of the output file
    the default is output.volumepkg and will be located in the 
    current working directory
"""
    sys.exit()

# get the scroll location from the cli arguements
# or by default from the current file
# if there arent any .tiff files then throw an error
if "-sp" in sys.argv:
    mypath = sys.argv[sys.argv.index("-sp") + 1]
elif "--slices-path" in sys.argv:
    mypath = sys.argv[sys.argv.index("--slices-path") + 1]
else:
    mypath = os.getcwd()

slices = sorted(
    filter(lambda f: f[-4::] == '.tif', 
        os.listdir(mypath))
    )

if len(slices) < 1:
    sys.exit("[Input Error] No Slices")

# parse arguements to see if we need to give the output file a certain name
# default is just output.volumepkg
if "-of" in sys.argv:
    outpath = sys.argv[sys.argv.index("-of") + 1]
elif "--output-file" in sys.argv:
    outpath = sys.argv[sys.argv.index("--output-file") + 1]
else:
    outpath = os.getcwd() + "/volume"

# get scan information
voxelSize = float(raw_input("Enter the voxel size in microns (um): "))
materialThickness = float(raw_input("Enter the estimated material thickness in microns (um): "))

# analyze the volume
for i in range(len(slices)):
    sys.stdout.write('\r')
    sys.stdout.write('Analyzing slice ' + str(i + 1) + '/' + str(len(slices)))
    # get width/height
    info = subprocess.check_output('vc_analyze "' + mypath+slices[i] + '" all', shell=True)
    thisWidth, thisHeight, thisDepth, thisMin, thisMax = info.split()
    
    # cast to the proper type
    thisWidth = int(thisWidth)
    thisHeight = int(thisHeight)
    thisDepth = int(thisDepth)
    thisMin = float(thisMin)
    thisMax = float(thisMax)

    # set intial values from first slice 
    if i == 0:
        width = thisWidth
        height = thisHeight
        depth = thisDepth
        volMin = thisMin
        volMax = thisMax
    else:
        # width, height, and depth must match slice 0
        if (thisWidth != width or
            thisHeight != height or
            thisDepth != depth):
                sys.exit('\nError: "' + mypath+slices[i] + '" does not match size/depth of volume!\n')
        # update volume min and max intensities if they're lower/higher respectively
        if thisMin < volMin:
            volMin = thisMin
        if thisMax > volMax:
            volMax = thisMax
    sys.stdout.flush()
sys.stdout.write('\n')

# scale min and max to 16-bit if 8-bit unsigned
if depth == 0:
    volMin = volMin*65535.00/255.00
    volMax = volMax*65535.00/255.00

# create the config options and save it to the config file
config_dict = {
    # volumepkg version
    "version": 2.0,
    # name
    "volumepkg name": outpath.split('/')[-1],

    # number of slices
    "number of slices": len(slices),
    "slice location": "slices/",

    # size and intensity info for volume
    "width": width,
    "height": height,
    "min": volMin,
    "max": volMax,
    "voxelsize": voxelSize,

    # material information
    "materialthickness": materialThickness
}

# create the folder structure
if os.path.isdir("tmp"):
    shutil.rmtree("tmp")
os.mkdir("tmp")
os.mkdir("tmp/slices")
os.mkdir("tmp/paths")
os.mkdir("tmp/surface_normals")

f = open('tmp/config.json','w')
f.write(json.dumps(config_dict))
f.close()

#copy the slices to the slices folder and reformat the names
for i in range(len(slices)):
    sys.stdout.write('\r')
    if depth != 2:
        sys.stdout.write('Conforming slice ' + str(i + 1) + '/' + str(len(slices)))
        os.system('vc_conform "' + mypath + slices[i] + '" tmp/slices/' + \
        '0'*(len(str(config_dict["number of slices"]))-len(str(i))) + \
        str(i) + '.tif')
    else:
        sys.stdout.write('Copying slice ' + str(i + 1) + '/' + str(len(slices)))
        os.system('cp "' + mypath + slices[i] + '" tmp/slices/' + \
        '0'*(len(str(config_dict["number of slices"]))-len(str(i))) + \
        str(i) + '.tif')
    sys.stdout.flush()
sys.stdout.write('\n\n')

#create the volumepkg file
if outpath.endswith(".volpkg") != True:
    outpath = outpath + ".volpkg"

# compress the volumepkg
# os.system('tar -zcvf ' + outpath + ' tmp/*')

# move the volumepkg as a folder
os.system('mv tmp "' + str(outpath) + '"')

#delete the tmp directory if it still exists
if os.path.isdir("tmp"):
    shutil.rmtree("tmp")

File: packager/packager.cpp
//
// Created by Seth Parker on 7/30/15.
//

#include "packager.h"

int main ( int argc, char* argv[]) {

    ///// Parse the command line options /////
    boost::filesystem::path slicesPath, volpkgPath;
    try {
        // All command line options
        boost::program_options::options_description options("Options");
        options.add_options()
                ("help,h", "Show this message")
                ("slices,s", boost::program_options::value<std::string>(), "Directory of input slice data")
                ("volpkg,v", boost::program_options::value<std::string>(), "Path for the output volume package");

        // parsedOptions will hold the values of all parsed options as a Map
        boost::program_options::variables_map parsedOptions;
        boost::program_options::store(boost::program_options::command_line_parser(argc, argv).options(options).run(), parsedOptions);
        boost::program_options::notify(parsedOptions);

        // Show the help message
        if (parsedOptions.count("help") || argc < 2) {
            std::cout << options << std::endl;
            return EXIT_SUCCESS;
        }

        // Get the input slices dir
        if (parsedOptions.count("slices")) {
            slicesPath = parsedOptions["slices"].as<std::string>();
        } else {
            std::cerr << "ERROR: Path to slices directory not supplied!" << std::endl;
            return EXIT_FAILURE;
        }

        // Get the output volpkg path
        if (parsedOptions.count("volpkg")) {
            volpkgPath = parsedOptions["volpkg"].as<std::string>();
        } else {
            std::cerr << "ERROR: Output Volume Package path not supplied!" << std::endl;
            return EXIT_FAILURE;
        }
    }
    catch(std::exception& e)
    {
        std::cerr << "ERROR: " << e.what() << std::endl;
        return EXIT_FAILURE;
    }

    ///// Setup /////
    // Check the extension and make sure the pkg doesn't already exist
    if ( volpkgPath.extension().string() != ".volpkg" ) volpkgPath.replace_extension(".volpkg");
    if (boost::filesystem::exists(volpkgPath)) {
        std::cerr << "ERROR: Volume package already exists at path specified." << std::endl;
        std::cerr << "This program does not currently allow for modification of existing volume packages." << std::endl;
        return EXIT_FAILURE;
    }

    // Generate an empty volpkg and save it to disk
    VolumePkg volpkg(volpkgPath.string(), VOLPKG_VERSION);
    volpkg.readOnly(false);
    volpkg.setMetadata("volumepkg name", volpkgPath.stem().string());
    volpkg.initialize();

    // Filter the slice path directory by extension and sort the vector of files
    std::vector<volcart::Slice> slices;
    if (boost::filesystem::exists(slicesPath) && boost::filesystem::is_directory(slicesPath)) {

        // Directory iterators
        boost::filesystem::directory_iterator dir_subfile(slicesPath);
        boost::filesystem::directory_iterator dir_end;

        // Filter out subfiles that aren't TIFs
        // To-Do: Handle other formats
        while (dir_subfile != dir_end) {
            std::string file_ext( boost::to_upper_copy<std::string>(dir_subfile->path().extension().string()) );
            if (is_regular_file(dir_subfile->path()) && (file_ext == ".TIF" || file_ext == ".TIFF")) {
                volcart::Slice temp;
                temp.path = *dir_subfile;
                slices.push_back(temp);
            }
            ++dir_subfile;
        }
    } else {
        std::cerr << "ERROR: Slices directory does not exist/is not a directory." << std::endl;
        std::cerr << "Please provide a directory of slice images." << std::endl;
        return EXIT_FAILURE;
    }
    if (slices.empty()) {
        std::cerr << "ERROR: No supported image files found in provided slices directory." << std::endl;
        return EXIT_FAILURE;
    }
    // Sort the Slices by their filenames
    std::sort(slices.begin(), slices.end(), SliceLess);


    ///// Analyze the slices /////
    bool vol_consistent = true;
    double vol_min, vol_max;
    for ( auto slice = slices.begin(); slice != slices.end(); ++slice ) {
        if (!slice->analyze()) continue; // skip if we can't analyze

        // Compare all slices to the properties of the first slice
        if (slice == slices.begin()) {
            vol_min = slice->min();
            vol_max = slice->max();
        }
        else {
            // Check for consistency first
            if (*slice != *slices.begin()) {
                vol_consistent = false;
                std::cerr << slice->path.filename() << " does not match the initial slice of the volume." << std::endl;
                continue;
            }

            // Update the volume's min and max
            if ( slice->min() < vol_min ) vol_min = slice->min();
            if ( slice->max() > vol_max ) vol_max = slice->max();
        }

    }
    if (!vol_consistent) {
        std::cerr << "ERROR: Slices in slice directory do not have matching properties (width/height/depth)." << std::endl;
        return EXIT_FAILURE;
    }

    ///// Add data to the volume package /////
    // Metadata
    volpkg.setMetadata("number of slices", slices.size());
    volpkg.setMetadata("width", slices.begin()->width());
    volpkg.setMetadata("height", slices.begin()->height());

    // Scale 8-bit min/max values
    // To-Do: Handle other bit depths
    if ( slices.begin()->depth() == 0 ) {
        vol_min = vol_min * 65535.00/255.00;
        vol_max = vol_max * 65535.00/255.00;
    }
    volpkg.setMetadata("min", vol_min);
    volpkg.setMetadata("max", vol_max);

    std::cout << "Saving slice images to " << volpkg.getPkgName() << "..." << std::endl;
    unsigned long count = 0;
    for ( auto slice = slices.begin(); slice != slices.end(); ++slice ) {
        volpkg.setSliceData(count, slice->image());
        ++count;
    }

    // Save final metadata changes to disk
    volpkg.saveMetadata();
    return EXIT_SUCCESS;
}